nixos/modules/services/web-servers/nginx/vhost-options.nix

```diff
@@ -212,6 +212,86 @@ with lib;
       description = "Path to root SSL certificate for stapling and client certificates.";
     };
 
+    # Since this is just an add_header, in principle it could be configured at
+    # the location level, but in practice I can't imagine that often being what
+    # people want, and they can always use `extraConfig` if necessary.
+    strictTransportSecurity = {
+      enable = mkOption {
+        type = types.bool;
+        default = false;
+        description = ''
+          Whether to enable HTTP Strict Transport Security (HSTS). This sends
+          the header "strict-transport-security: max-age=63072000" on all
+          responses, which tells clients to refuse to use an insecure HTTP
+          connection for this host for the next 63072000 seconds (approximately
+          2 years), or for the duration specified by
+          `strictTransportSecurity.seconds` if set. This helps protects against
+          man-in-the-middle attacks that e.g. block HTTPS connections in the
+          hope that the client will fall back to an insecure HTTP connection,
+          which can be intercepted and modified.
+
+          The strict-transport-security header is (sent but) ignored on HTTP
+          connections, so this setting will only take effect if your website
+          actually is available over HTTPS.
+
+          HSTS only protects clients who have managed to connect at least once
+          without the man in the middle in the last two years: see also
+          `strictTransportSecurity.preload` for how to protect even the very
+          first connection.
+
+          HSTS is recommended in cases where you can reliably serve HTTPS, but
+          very annoying if you temporarily set it up but then want to use HTTP
+          again, so it is not enabled by default.
+        '';
+      };
+
+      seconds = mkOption {
+        type = types.int;
+        default = 63072000;
+        description = ''
+          If HTTP Strict Transport Security (HSTS) is enabled by
+          `strictTransportSecurity.enable`, this is how long the browser is
+          instructed to retain the setting, in seconds. 2 years (63072000 seconds)
+          is recommended, but if you're introducing HSTS for the first time on an
+          existing site you may want to ramp up the value gradually, or set it to
+          0 to tell clients to drop their HSTS config for this domain.
+        '';
+      };
+
+      includeSubdomains = mkOption {
+        type = types.bool;
+        default = true;
```

Comment:
I am not sure this should be on by default, it is a footgun unless you are really prepared. Lots of cases of random sub‐subdomains with internal test deployments or services that aren’t ready for it.

(Haven’t done a thorough review of the rest of the PR yet.)

Reply:
sure, done :)

tbh my attitude was "enableACME is so easy these days" but I guess non-public uses have the potential to be trickier

Reply:
Yeah, the problem is that it‘s easy if you have one single NixOS machine serving your entire domain hierarchy, and it‘s easy if you have a trillion machines that are all behind consistent HTTPS frontend stuff proxying everything to your highly‐structured web scale enterprise backends, but there‘s a wide range of setups between those where you just don‘t have complete visibility and control into everything going on under `*.mycorp.com`. Intranet, legacy SaaS, a random appliance you can‘t update, … there’s a big gap between “this web server on this domain can reliably do HTTPS” and “absolutely nothing under this hierarchy exists that cannot reliably do HTTPS in every single context”, and with HSTS it’s easy to screw up painfully. Of course people should set it if they can.

Will try to take a more detailed look at this PR in the next few days if nobody beats me to it, feel free to ping if I forget.